modulemd/v2/modulemd-profile.c

```diff
@@ -0,0 +1,461 @@
+    }
+}
+
+
+static void
+modulemd_profile_class_init (ModulemdProfileClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->finalize = modulemd_profile_finalize;
+  object_class->get_property = modulemd_profile_get_property;
+  object_class->set_property = modulemd_profile_set_property;
+
+  properties[PROP_NAME] = g_param_spec_string (
+    "name",
+    "Name",
+    "The name of this profile.",
+    P_DEFAULT_STRING,
+    G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS | G_PARAM_CONSTRUCT_ONLY);
+  properties[PROP_DESCRIPTION] =
+    g_param_spec_string ("description",
+                         "Description",
+                         "The untranslated description of this profile",
+                         P_DEFAULT_STRING,
+                         G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS);
+
+  g_object_class_install_properties (object_class, N_PROPS, properties);
+}
+
+
+static void
+modulemd_profile_init (ModulemdProfile *self)
+{
+  self->rpms = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+}
+
+
+/* === YAML Functions === */
+
+ModulemdProfile *
+modulemd_profile_parse_yaml (yaml_parser_t *parser, GError **error)
+{
+  MODULEMD_INIT_TRACE ();
+  MMD_INIT_YAML_EVENT (event);
+  gboolean done = FALSE;
+  gboolean in_map = FALSE;
+  gboolean in_rpms = FALSE;
+  g_autofree gchar *value;
+  g_autoptr (ModulemdProfile) p = NULL;
+  g_autoptr (GError) nested_error = NULL;
+
+  g_return_val_if_fail (error == NULL || *error == NULL, FALSE);
+
+  /* Read the profile name */
+  YAML_PARSER_PARSE_WITH_EXIT (parser, &event, error);
+  if (event.type != YAML_SCALAR_EVENT)
+    {
+      MMD_YAML_ERROR_EVENT_EXIT (error, event, "Missing profile name");
+    }
+  p = modulemd_profile_new ((const gchar *)event.data.scalar.value);
+  yaml_event_delete (&event);
+
+  /* Read in additional attributes */
+  while (!done)
+    {
+      YAML_PARSER_PARSE_WITH_EXIT (parser, &event, error);
+
+      switch (event.type)
+        {
+        case YAML_MAPPING_START_EVENT: in_map = TRUE; break;
+
+        case YAML_MAPPING_END_EVENT:
+          in_map = FALSE;
+          done = TRUE;
+          break;
+
+        case YAML_SEQUENCE_START_EVENT:
+          if (!in_rpms)
+            {
+              MMD_YAML_ERROR_EVENT_EXIT (
+                error, event, "Unexpected list in profile");
+            }
+          if (!in_map)
+            {
+              MMD_YAML_ERROR_EVENT_EXIT (
+                error, event, "Missing mapping in profile entry");
+            }
+          in_rpms = TRUE;
+          break;
+
+        case YAML_SEQUENCE_END_EVENT:
+          if (!in_rpms)
+            MMD_YAML_ERROR_EVENT_EXIT (error, event, "Unexpected end of list");
+          in_rpms = FALSE;
+          break;
+
+        case YAML_SCALAR_EVENT:
+          if (!in_map)
+            MMD_YAML_ERROR_EVENT_EXIT (
+              error, event, "Missing mapping in profile entry");
+          if (in_rpms)
```

Comment:
Since this is going to come up frequently in other places (there are several such lists in ModuleStream), could you instead add a common routine to read in all strings in a YAML sequence into a passed GHashTable? That will also simplify the logic at this layer, rather than needing to maintain in_rpms as well.